File: app/src/main/java/jp/kentan/student_portal_plus/ui/SettingsActivity.java
package jp.kentan.student_portal_plus.ui;


import android.app.FragmentManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.CheckBoxPreference;
import android.preference.ListPreference;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.SwitchPreference;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Toast;

import jp.kentan.student_portal_plus.R;
import jp.kentan.student_portal_plus.data.DatabaseProvider;
import jp.kentan.student_portal_plus.data.PortalDataProvider;
import jp.kentan.student_portal_plus.data.component.LectureCancellation;
import jp.kentan.student_portal_plus.data.component.LectureInformation;
import jp.kentan.student_portal_plus.data.component.MyClass;
import jp.kentan.student_portal_plus.data.component.News;
import jp.kentan.student_portal_plus.notification.NotificationScheduler;
import jp.kentan.student_portal_plus.notification.NotificationService;
import jp.kentan.student_portal_plus.ui.widget.MyClassSamplePreference;

public class SettingsActivity extends AppCompatActivity {

    private static SettingsActivity sActivity;

    private static SharedPreferences sPreferenceCommon, sPreferenceNotification;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTitle(getString(R.string.title_activity_settings));

        /*
        Fragment Initialize
         */
        FragmentManager fm = getFragmentManager();
        for(int i = 0; i < fm.getBackStackEntryCount(); ++i) {
            fm.popBackStack();
        }

        getFragmentManager().beginTransaction().replace(android.R.id.content, new PreferencesFragment()).commit();
        setupActionBar();

        sActivity = this;

        sPreferenceCommon = sActivity.getSharedPreferences("common", MODE_PRIVATE);
        sPreferenceNotification = sActivity.getSharedPreferences("notification", MODE_PRIVATE);
    }

    private void setupActionBar() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        onBackPressed();
        return true;
    }

    @Override
    public void onBackPressed() {
        final int stackCnt = getFragmentManager().getBackStackEntryCount();
        if (stackCnt > 0) {
            getFragmentManager().popBackStack();

            if (stackCnt <= 1) {
                setTitle(getString(R.string.title_activity_settings));
            }
            return;
        }

        super.onBackPressed();
    }


    public static class PreferencesFragment extends PreferenceFragment implements Preference.OnPreferenceClickListener, SharedPreferences.OnSharedPreferenceChangeListener {

        private Preference mNotifyContents;
        private ListPreference mAutoFetchInterval, mLimitOfLatestInfo;
        private CheckBoxPreference mNotifyWithVibrate, mNotifyWithLed;

        private NotificationScheduler mNotificationScheduler;

        private static Toast sToast;
        private int countVersionClick = 5;
        private final static String[] SECRET_TEXT = {"nyan~ > \uD83D\uDC31", "(´・ω・｀)", "have a nice day \uD83C\uDF08", "thanks <3"};


        @Override
        public void onCreate(Bundle saveInstanceState) {
            super.onCreate(saveInstanceState);
            addPreferencesFromResource(R.xml.preferences);

            mNotificationScheduler = new NotificationScheduler(getActivity());


            /*
            Set default values
             */
            mNotifyContents = getPreferenceScreen().findPreference("pref_key_notify_contents");
            mAutoFetchInterval = (ListPreference) getPreferenceScreen().findPreference("pref_key_auto_fetch_interval");
            mLimitOfLatestInfo = (ListPreference) getPreferenceScreen().findPreference("pref_key_limit_of_latest_info");
            mNotifyWithVibrate = (CheckBoxPreference) getPreferenceScreen().findPreference("pref_key_notify_with_vibrate");
            mNotifyWithLed = (CheckBoxPreference) getPreferenceScreen().findPreference("pref_key_notify_with_led");

            final boolean isAutoFetch = sPreferenceCommon.getBoolean("auto_fetch", true);
            mNotifyContents.setEnabled(isAutoFetch);
            mAutoFetchInterval.setEnabled(isAutoFetch);
            mNotifyWithVibrate.setEnabled(isAutoFetch);
            mNotifyWithLed.setEnabled(isAutoFetch);


            /*
            Set EventListener
             */
            mNotifyContents.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                public boolean onPreferenceClick(Preference preference) {
                    getFragmentManager().beginTransaction().replace(android.R.id.content, new NotifyContentsPreferencesFragment()).addToBackStack(null).commit();
                    return true;
                }
            });

            getPreferenceScreen().findPreference("pref_key_select_update_data").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                public boolean onPreferenceClick(Preference preference) {
                    getFragmentManager().beginTransaction().replace(android.R.id.content, new SelectUpdateContentsPreferencesFragment()).addToBackStack(null).commit();
                    return true;
                }
            });
            getPreferenceScreen().findPreference("pref_key_my_class_threshold").setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                public boolean onPreferenceClick(Preference preference) {
                    getFragmentManager().beginTransaction().replace(android.R.id.content, new MyClassThresholdPreferencesFragment()).addToBackStack(null).commit();
                    return true;
                }
            });

            getPreferenceScreen().findPreference("pref_key_delete_all_data").setOnPreferenceClickListener(this);
            getPreferenceScreen().findPreference("pref_key_terms").setOnPreferenceClickListener(this);
            getPreferenceScreen().findPreference("pref_key_oss_license").setOnPreferenceClickListener(this);
            getPreferenceScreen().findPreference("pref_key_version").setOnPreferenceClickListener(this);


            updatePrefSummary();
        }

        private void updatePrefSummary() {
            updateLastLoginDate();

            final int interval = sPreferenceCommon.getInt("auto_fetch_interval_m", 60);
            String autoFetchSummary;

            if (interval >= 60) {
                autoFetchSummary = Integer.toString(interval / 60) + "時間毎に更新する";
            } else {
                autoFetchSummary = Integer.toString(interval) + "分毎に更新する";
            }

            mAutoFetchInterval.setSummary(autoFetchSummary);
            mLimitOfLatestInfo.setSummary(sPreferenceCommon.getInt("limit_of_latest_info", 100) + "件 (古い情報は自動的に消去されます)");
        }

        private void updateLastLoginDate(){
            String strLastLoginDate = sActivity.getSharedPreferences("IDP", Context.MODE_PRIVATE).getString("last_login_date", "未ログイン");
            getPreferenceScreen().findPreference("pref_key_idp_last_login_date").setSummary(strLastLoginDate);
        }

        @Override
        public boolean onPreferenceClick(Preference pref) {
            switch (pref.getKey()) {
                case "pref_key_delete_all_data":
                    AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
                    builder.setIcon(R.drawable.ic_warning);
                    builder.setTitle("ポータルデータ消去");
                    builder.setMessage(getString(R.string.msg_warn_reset));
                    builder.setNegativeButton("いいえ", null);

                    builder.setPositiveButton("はい", new DialogInterface.OnClickListener() {

                        public void onClick(DialogInterface dialog, int id) {
                            Toast.makeText(getActivity(), DatabaseProvider.deleteAll() + "件のデータを消去しました", Toast.LENGTH_LONG).show();
                        }

                    });

                    builder.show();
                    break;
                case "pref_key_terms":
                    AlertDialog.Builder dialogTerms = new AlertDialog.Builder(sActivity);
                    dialogTerms.setTitle(getString(R.string.pref_title_terms));

                    final WebView webTerms = new WebView(sActivity);
                    webTerms.getSettings().setCacheMode(WebSettings.LOAD_NO_CACHE);
                    webTerms.getSettings().setAppCacheEnabled(false);
                    webTerms.loadUrl(getString(R.string.url_terms));
                    webTerms.setWebViewClient(new WebViewClient() {
                        @Override
                        public void onPageFinished(WebView view, String url) {
                            //オンライン取得に失敗
                            if (!webTerms.getTitle().contains(getString(R.string.title_terms))) {
                                webTerms.loadUrl(getString(R.string.url_terms_local));
                            }
                        }
                    });
                    dialogTerms.setView(webTerms);
                    dialogTerms.setNegativeButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int id) {
                            dialog.dismiss();
                        }
                    });
                    dialogTerms.show();
                    break;
                case "pref_key_oss_license":
                    AlertDialog.Builder dialogOssLicenses = new AlertDialog.Builder(sActivity);
                    dialogOssLicenses.setTitle(getString(R.string.pref_title_oss_license));

                    WebView webOssLicenses = new WebView(sActivity);
                    webOssLicenses.loadUrl("file:///android_asset/licenses.html");
                    webOssLicenses.setWebViewClient(new WebViewClient());
                    dialogOssLicenses.setView(webOssLicenses);
                    dialogOssLicenses.setNegativeButton("OK", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int id) {
                            dialog.dismiss();
                        }
                    });
                    dialogOssLicenses.show();
                    break;
                case "pref_key_version":
                    final String msg;
                    if(countVersionClick < 1){
                        countVersionClick = 5;
                        msg = SECRET_TEXT[(int)(Math.random() * SECRET_TEXT.length)];
                    }else{
                        msg = Integer.toString(countVersionClick--);
                    }

                    if(sToast != null) sToast.cancel();

                    sToast = Toast.makeText(getActivity(), msg, Toast.LENGTH_SHORT);
                    sToast.show();
                    break;
            }
            return true;
        }

        @Override
        public void onSharedPreferenceChanged(SharedPreferences pref, String key) {
            final SharedPreferences.Editor editorCommon = sPreferenceCommon.edit();
            final SharedPreferences.Editor editorNotification = sPreferenceNotification.edit();

            switch (key) {
                case "pref_key_auto_fetch":
                    boolean isFetch = pref.getBoolean(key, true);

                    editorCommon.putBoolean("auto_fetch", isFetch);
                    mNotifyContents.setEnabled(isFetch);
                    mAutoFetchInterval.setEnabled(isFetch);
                    mNotifyWithVibrate.setEnabled(isFetch);
                    mNotifyWithLed.setEnabled(isFetch);

                    if (isFetch) {
                        mNotificationScheduler.schedule();
                    } else {
                        mNotificationScheduler.cancel();
                    }
                    break;
                case "pref_key_auto_fetch_interval":
                    final int interval = Integer.parseInt(mAutoFetchInterval.getValue());
                    String autoFetchSummary;

                    if (interval >= 60) {
                        autoFetchSummary = Integer.toString(interval / 60) + "時間毎に更新する";
                    } else {
                        autoFetchSummary = Integer.toString(interval) + "分毎に更新する";
                    }

                    editorCommon.putInt("auto_fetch_interval_m", interval);
                    mAutoFetchInterval.setSummary(autoFetchSummary);

                    mNotificationScheduler.schedule(interval);
                    break;
                case "pref_key_notify_with_vibrate":
                    editorNotification.putBoolean("vibrate", pref.getBoolean(key, true));
                    break;
                case "pref_key_notify_with_led":
                    editorNotification.putBoolean("led", pref.getBoolean(key, true));
                    break;
                case "pref_key_limit_of_latest_info":
                    String limit = pref.getString(key, "90");

                    editorCommon.putInt("limit_of_latest_info", Integer.parseInt(limit));
                    mLimitOfLatestInfo.setSummary(limit + "件 (古い情報は自動的に消去されます)");
                    break;
                case "pref_key_pdf_open_with_gdocs":
                    editorCommon.putBoolean("pdf_open_with_gdocs", pref.getBoolean(key, true));
                    break;
                case "pref_key_detail_err_msg":
                    editorCommon.putBoolean("detail_err_msg", pref.getBoolean(key, false));
                    break;
            }

            editorCommon.apply();
            editorNotification.apply();
        }

        @Override
        public void onResume() {
            super.onResume();
            getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
            updateLastLoginDate();
        }

        @Override
        public void onPause() {
            super.onPause();
            getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        }
    }


    public static class NotifyContentsPreferencesFragment extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener {

        private final static String[] NOTIFY_TYPE = {"通知しない", "すべて", "受講科目・類似科目"};

        private ListPreference mNotifyLectureInfo, mNotifyCancelInfo, mNotifyLatestInfo;

        @Override
        public void onCreate(Bundle saveInstanceState) {
            super.onCreate(saveInstanceState);
            addPreferencesFromResource(R.xml.pref_notify_contents);

            mNotifyLectureInfo = (ListPreference) findPreference("pref_key_notify_type_lecture_info");
            mNotifyCancelInfo = (ListPreference) findPreference("pref_key_notify_type_cancel_info");
            mNotifyLatestInfo = (ListPreference) findPreference("pref_key_notify_type_latest_info");

            getActivity().setTitle("新着通知");

            updatePrefSummary();
        }

        private void updatePrefSummary() {
            mNotifyLectureInfo.setSummary(NOTIFY_TYPE[sPreferenceNotification.getInt("type_lecture_info", NotificationService.NOTIFY_ALL) + 1]);
            mNotifyCancelInfo.setSummary( NOTIFY_TYPE[sPreferenceNotification.getInt("type_cancel_info", NotificationService.NOTIFY_ALL) + 1]);
            mNotifyLatestInfo.setSummary( NOTIFY_TYPE[sPreferenceNotification.getInt("type_latest_info", NotificationService.NOTIFY_ALL) + 1]);
        }

        @Override
        public void onSharedPreferenceChanged(SharedPreferences pref, String key) {
            final SharedPreferences.Editor editor = sPreferenceNotification.edit();
            final int index = Integer.parseInt(pref.getString(key, ""));

            switch (key) {
                case "pref_key_notify_type_lecture_info":

                    mNotifyLectureInfo.setSummary(NOTIFY_TYPE[index + 1]);
                    editor.putInt("type_lecture_info", index);
                    break;
                case "pref_key_notify_type_cancel_info":

                    mNotifyCancelInfo.setSummary(NOTIFY_TYPE[index + 1]);
                    editor.putInt("type_cancel_info", index);
                    break;
                case "pref_key_notify_type_latest_info":

                    mNotifyLatestInfo.setSummary(NOTIFY_TYPE[index + 1]);
                    editor.putInt("type_latest_info", index);
                    break;
            }

            editor.apply();
        }

        @Override
        public void onResume() {
            super.onResume();
            getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        }

        @Override
        public void onPause() {
            getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
            super.onPause();
        }
    }


    public static class SelectUpdateContentsPreferencesFragment extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener {

        private final static String LAST_DATE_HEAD = "最終更新: ";

        private SharedPreferences mPreferences;

        private CheckBoxPreference mMyClass, mLectureInfo, mCancelInfo, mLatestInfo;

        @Override
        public void onCreate(Bundle saveInstanceState) {
            super.onCreate(saveInstanceState);
            addPreferencesFromResource(R.xml.pref_select_update_contents);

            mPreferences = getActivity().getSharedPreferences("update_content", MODE_PRIVATE);

            getActivity().setTitle("ポータルデータ");

            SwitchPreference mAll = (SwitchPreference) findPreference("pref_key_update_all");
            mMyClass = (CheckBoxPreference) findPreference("pref_key_update_my_class");
            mLectureInfo = (CheckBoxPreference) findPreference("pref_key_update_lecture_info");
            mCancelInfo = (CheckBoxPreference) findPreference("pref_key_update_cancel_info");
            mLatestInfo = (CheckBoxPreference) findPreference("pref_key_update_latest_info");

            mMyClass.setChecked(mPreferences.getBoolean(MyClass.KEY, true));
            mLectureInfo.setChecked(mPreferences.getBoolean(LectureInformation.KEY, true));
            mCancelInfo.setChecked(mPreferences.getBoolean(LectureCancellation.KEY, true));
            mLatestInfo.setChecked(mPreferences.getBoolean(News.KEY, true));

            if (mPreferences.getBoolean("all", true)) {
                mAll.setChecked(true);
                setEnabledAllCheckBox(false);
            } else {
                mAll.setChecked(false);
            }

            updatePrefSummary();
        }

        private void updatePrefSummary() {
            mMyClass.setSummary(    LAST_DATE_HEAD + mPreferences.getString(MyClass.KEY_DATE, "未取得"));
            mLectureInfo.setSummary(LAST_DATE_HEAD + mPreferences.getString(LectureInformation.KEY_DATE, "未取得"));
            mCancelInfo.setSummary( LAST_DATE_HEAD + mPreferences.getString(LectureCancellation.KEY_DATE, "未取得"));
            mLatestInfo.setSummary( LAST_DATE_HEAD + mPreferences.getString(News.KEY_DATE, "未取得"));
        }

        private void setEnabledAllCheckBox(boolean enabled) {
            mMyClass.setEnabled(enabled);
            mLectureInfo.setEnabled(enabled);
            mCancelInfo.setEnabled(enabled);
            mLatestInfo.setEnabled(enabled);
        }

        @Override
        public void onSharedPreferenceChanged(SharedPreferences pref, String key) {
            final SharedPreferences.Editor editor = mPreferences.edit();

            switch (key) {
                case "pref_key_update_all":
                    final boolean isAll = pref.getBoolean(key, true);

                    editor.putBoolean("all", isAll);
                    setEnabledAllCheckBox(!isAll);
                    break;
                case "pref_key_update_my_class":
                    editor.putBoolean(MyClass.KEY, pref.getBoolean(key, true));
                    break;
                case "pref_key_update_lecture_info":
                    editor.putBoolean(LectureInformation.KEY, pref.getBoolean(key, true));
                    break;
                case "pref_key_update_cancel_info":
                    editor.putBoolean(LectureCancellation.KEY, pref.getBoolean(key, true));
                    break;
                case "pref_key_update_latest_info":
                    editor.putBoolean(News.KEY, pref.getBoolean(key, true));
                    break;
            }

            editor.apply();
        }

        @Override
        public void onResume() {
            super.onResume();
            getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        }

        @Override
        public void onPause() {
            super.onPause();
            getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        }
    }


    public static class MyClassThresholdPreferencesFragment extends PreferenceFragment implements SharedPreferences.OnSharedPreferenceChangeListener {

        private ListPreference mMyClassThreshold;
        private MyClassSamplePreference mMyClassSample;


        @Override
        public void onCreate(Bundle saveInstanceState) {
            super.onCreate(saveInstanceState);
            addPreferencesFromResource(R.xml.pref_my_class_threshold);

            getActivity().setTitle("ポータルデータ");

            mMyClassThreshold = (ListPreference)findPreference("pref_key_my_class_threshold");
            mMyClassSample = (MyClassSamplePreference)findPreference("pref_key_my_class_threshold_sample");

            updatePrefSummary(sPreferenceCommon.getFloat("my_class_threshold", 0.8f));
        }

        private void updatePrefSummary(final float threshold) {
            final int summary = (int)(threshold*100.0f);

            mMyClassThreshold.setSummary(summary + "%%" + ((summary < 100) ? "以上" : ""));
            mMyClassSample.setThreshold(threshold);
        }

        @Override
        public void onSharedPreferenceChanged(SharedPreferences pref, String key) {
            if(key.equals("pref_key_my_class_threshold")){
                final SharedPreferences.Editor editor = sPreferenceCommon.edit();
                float threshold = Float.parseFloat(pref.getString(key, "80")) / 100.0f;

                PortalDataProvider.setMyClassThreshold(threshold);

                editor.putFloat("my_class_threshold", threshold);
                editor.apply();

                updatePrefSummary(threshold);
            }
        }

        @Override
        public void onResume() {
            super.onResume();
            getPreferenceManager().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
        }

        @Override
        public void onPause() {
            super.onPause();
            getPreferenceManager().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
        }
    }
}

File: app/src/main/java/jp/kentan/studentportalplus/ui/SettingsActivity.kt
package jp.kentan.studentportalplus.ui

import android.app.FragmentTransaction
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.ListPreference
import android.preference.Preference
import android.preference.PreferenceFragment
import android.preference.PreferenceScreen
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem
import dagger.android.AndroidInjection
import jp.kentan.studentportalplus.BuildConfig
import jp.kentan.studentportalplus.R
import jp.kentan.studentportalplus.data.PortalRepository
import jp.kentan.studentportalplus.data.component.NotifyType
import jp.kentan.studentportalplus.notification.NotificationController
import jp.kentan.studentportalplus.notification.SyncScheduler
import jp.kentan.studentportalplus.ui.span.CustomTitle
import jp.kentan.studentportalplus.ui.widget.MyClassThresholdSamplePreference
import jp.kentan.studentportalplus.util.enabledSync
import jp.kentan.studentportalplus.util.getSyncIntervalMinutes
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.launch
import org.jetbrains.anko.coroutines.experimental.bg
import org.jetbrains.anko.defaultSharedPreferences
import org.jetbrains.anko.longToast
import org.jetbrains.anko.startActivity
import javax.inject.Inject


class SettingsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        title = CustomTitle(this, getString(R.string.title_activity_settings))

        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        if (savedInstanceState == null) {
            fragmentManager
                    .beginTransaction()
                    .add(android.R.id.content, PreferencesFragment())
                    .commit()
        }

        NotificationController.setupChannel(this)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        onBackPressed()
        return true
    }

    override fun onBackPressed() {
        val count = fragmentManager.backStackEntryCount

        if (count > 0) {
            fragmentManager.popBackStackImmediate()
            return
        }

        super.onBackPressed()
    }

    class PreferencesFragment : PreferenceFragment(), SharedPreferences.OnSharedPreferenceChangeListener {

        @Inject
        lateinit var portalRepository: PortalRepository

        private lateinit var shibbolethLastLoginDate: Preference
        private lateinit var syncInterval: ListPreference
        private lateinit var notifyContents: Preference
        private var notifyVibration: Preference? = null
        private var notifyLed: Preference? = null

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            addPreferencesFromResource(R.xml.preferences)

            AndroidInjection.inject(this)

            val screen = preferenceScreen

            shibbolethLastLoginDate = screen.findPreference("shibboleth_last_login_date")
            syncInterval = screen.findPreference("sync_interval") as ListPreference
            notifyContents = screen.findPreference("notify_contents")
            notifyVibration = screen.findPreference("enabled_notification_vibration")
            notifyLed = screen.findPreference("enabled_notification_led")

            screen.findPreference("notification_settings")?.setOnPreferenceClickListener {
                val intent = Intent("android.settings.CHANNEL_NOTIFICATION_SETTINGS")
                intent.putExtra("android.provider.extra.APP_PACKAGE", activity.packageName)
                intent.putExtra("android.provider.extra.CHANNEL_ID", NotificationController.NEWLY_CHANNEL_ID)
                startActivity(intent)

                return@setOnPreferenceClickListener true
            }

            notifyContents.setOnPreferenceClickListener {
                fragmentManager
                        .beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(android.R.id.content, NotifyContentsFragment())
                        .addToBackStack(null)
                        .commit()

                return@setOnPreferenceClickListener true
            }

            screen.findPreference("my_class_threshold").setOnPreferenceClickListener {
                fragmentManager
                        .beginTransaction()
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                        .replace(android.R.id.content, MyClassThresholdFragment())
                        .addToBackStack(null)
                        .commit()

                return@setOnPreferenceClickListener true
            }


            val enabled = defaultSharedPreferences.enabledSync()

            syncInterval.isEnabled = enabled
            notifyContents.isEnabled = enabled
            notifyVibration?.isEnabled = enabled
            notifyLed?.isEnabled = enabled

            setupSummary(screen)
        }

        override fun onPreferenceTreeClick(preferenceScreen: PreferenceScreen, pref: Preference): Boolean {
            when (pref.key) {
                "reset" -> {
                    AlertDialog.Builder(activity)
                            .setIcon(R.drawable.ic_warning)
                            .setTitle("ポータルデータ消去")
                            .setMessage(R.string.msg_warn_reset)
                            .setPositiveButton(R.string.action_yes) { _, _ ->
                                bg {
                                    portalRepository.deleteAll { success ->
                                        launch(UI) {
                                            if (success) {
                                                longToast("すべてのポータルデータを消去しました")
                                            } else {
                                                longToast("消去に失敗しました")
                                            }
                                        }
                                    }
                                }
                            }
                            .setNegativeButton(R.string.action_no, null)
                            .show()
                }
                "share" -> {
                    val intent = Intent(Intent.ACTION_SEND)
                    intent.type = "text/plain"
                    intent.putExtra(Intent.EXTRA_TEXT, getString(R.string.text_share_app))
                    startActivity(intent)
                }
                "terms" -> {
                    startActivity<WebViewActivity>("title" to "Terms", "url" to getString(R.string.url_terms))
                }
                "oss_license" -> {
                    startActivity<WebViewActivity>("title" to "Licenses", "url" to getString(R.string.url_licenses))
                }
            }
            return super.onPreferenceTreeClick(preferenceScreen, pref)
        }

        override fun onSharedPreferenceChanged(pref: SharedPreferences, key: String) {
            when (key) {
                "shibboleth_last_login_date" -> {
                    shibbolethLastLoginDate.summary = pref.getString("shibboleth_last_login_date", "なし")
                }
                "enabled_sync" -> {
                    val enable = pref.getBoolean("enabled_sync", true)

                    syncInterval.isEnabled = enable
                    notifyContents.isEnabled = enable
                    notifyVibration?.isEnabled = enable
                    notifyLed?.isEnabled = enable

                    if (enable) {
                        SyncScheduler.schedule(activity)
                    } else {
                        SyncScheduler.cancel()
                    }
                }
                "sync_interval" -> {
                    val interval = pref.getSyncIntervalMinutes()

                    syncInterval.summary = if (interval >= 60) {
                        "${interval / 60}時間毎に更新する"
                    } else {
                        "${interval}分毎に更新する"
                    }

                    SyncScheduler.schedule(activity)
                }
            }
        }

        override fun onResume() {
            super.onResume()
            defaultSharedPreferences.registerOnSharedPreferenceChangeListener(this)
        }

        override fun onPause() {
            super.onPause()
            defaultSharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        }

        private fun setupSummary(screen: PreferenceScreen) {
            val pref = defaultSharedPreferences

            screen.findPreference("shibboleth_last_login_date").summary = pref.getString("shibboleth_last_login_date", "なし")
            screen.findPreference("version").summary = BuildConfig.VERSION_NAME

            val interval = pref.getSyncIntervalMinutes()
            syncInterval.summary = if (interval >= 60) {
                "${interval / 60}時間毎に更新する"
            } else {
                "${interval}分毎に更新する"
            }
        }
    }

    class NotifyContentsFragment : PreferenceFragment(), SharedPreferences.OnSharedPreferenceChangeListener {

        private lateinit var lectureInfo: Preference
        private lateinit var lectureCancel: Preference
        private lateinit var notice: Preference

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            addPreferencesFromResource(R.xml.pref_notify_contents)

            val screen = preferenceScreen
            val pref = defaultSharedPreferences

            lectureInfo = screen.findPreference("notify_type_lecture_info")
            lectureCancel = screen.findPreference("notify_type_lecture_cancel")
            notice = screen.findPreference("notify_type_notice")

            lectureInfo.summary = NotifyType.valueOf(pref.getString("notify_type_lecture_info", NotifyType.ALL.name)).displayName
            lectureCancel.summary = NotifyType.valueOf(pref.getString("notify_type_lecture_cancel", NotifyType.ALL.name)).displayName
            notice.summary = NotifyType.valueOf(pref.getString("notify_type_notice", NotifyType.ALL.name)).displayName

        }

        override fun onSharedPreferenceChanged(pref: SharedPreferences, key: String) {

            val preference = when (key) {
                "notify_type_lecture_info" -> lectureInfo
                "notify_type_lecture_cancel" -> lectureCancel
                "notify_type_notice" -> notice
                else -> return
            }

            preference.summary = NotifyType.valueOf(pref.getString(key, NotifyType.ALL.name)).displayName
        }

        override fun onResume() {
            super.onResume()
            defaultSharedPreferences.registerOnSharedPreferenceChangeListener(this)
        }

        override fun onPause() {
            super.onPause()
            defaultSharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
        }
    }


    class MyClassThresholdFragment : PreferenceFragment() {

        private lateinit var threshold: Preference
        private lateinit var sample: MyClassThresholdSamplePreference

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            addPreferencesFromResource(R.xml.pref_my_class_threshold)

            val screen = preferenceScreen

            threshold = screen.findPreference("my_class_threshold")
            sample    = screen.findPreference("my_class_threshold_sample") as MyClassThresholdSamplePreference

            val currentPercent = defaultSharedPreferences.getString("my_class_threshold", "80").toIntOrNull() ?: 80

            threshold.summary = if (currentPercent < 100) "$currentPercent%%以上" else "100%%"
            sample.updateThreshold(currentPercent)

            threshold.setOnPreferenceChangeListener { _, newValue ->
                val percent = newValue.toString().toIntOrNull() ?: 80

                threshold.summary = if (percent < 100) "$newValue%%以上" else "100%%"
                sample.updateThreshold(percent)

                return@setOnPreferenceChangeListener true
            }
        }
    }
}
